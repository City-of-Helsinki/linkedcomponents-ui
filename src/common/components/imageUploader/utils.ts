/* eslint-disable @typescript-eslint/no-explicit-any */
import imageCompression from 'browser-image-compression';
import { PixelCrop } from 'react-image-crop';

import {
  COMPRESSABLE_IMAGE_TYPES,
  MAX_IMAGE_SIZE_MB,
  MAX_IMAGE_WIDTH,
} from '../../../constants';
import isTestEnv from '../../../utils/isTestEnv';

const TO_RADIANS = Math.PI / 180;

const toBlob = (
  canvas: HTMLCanvasElement,
  type?: string,
  quality?: any
): Promise<Blob | null> =>
  new Promise((resolve) => {
    canvas.toBlob(resolve, type, quality);
  });

const canvasPreview = (
  image: HTMLImageElement,
  canvas: HTMLCanvasElement,
  crop: PixelCrop,
  scale = 1,
  rotate = 0
) => {
  const ctx = canvas.getContext('2d');

  if (!ctx) {
    throw new Error('No 2d context');
  }

  const scaleX = image.naturalWidth / image.width;
  const scaleY = image.naturalHeight / image.height;
  // devicePixelRatio slightly increases sharpness on retina devices
  // at the expense of slightly slower render times and needing to
  // size the image back down if you want to download/upload and be
  // true to the images natural size.
  const pixelRatio = window.devicePixelRatio;
  // const pixelRatio = 1

  canvas.width = Math.floor(crop.width * scaleX * pixelRatio);
  canvas.height = Math.floor(crop.height * scaleY * pixelRatio);

  ctx.scale(pixelRatio, pixelRatio);
  ctx.imageSmoothingQuality = 'high';

  const cropX = crop.x * scaleX;
  const cropY = crop.y * scaleY;

  const rotateRads = rotate * TO_RADIANS;
  const centerX = image.naturalWidth / 2;
  const centerY = image.naturalHeight / 2;

  ctx.save();

  // 5) Move the crop origin to the canvas origin (0,0)
  ctx.translate(-cropX, -cropY);
  // 4) Move the origin to the center of the original position
  ctx.translate(centerX, centerY);
  // 3) Rotate around the origin
  ctx.rotate(rotateRads);
  // 2) Scale the image
  ctx.scale(scale, scale);
  // 1) Move the center of the image to the origin (0,0)
  ctx.translate(-centerX, -centerY);
  ctx.drawImage(
    image,
    0,
    0,
    image.naturalWidth,
    image.naturalHeight,
    0,
    0,
    image.naturalWidth,
    image.naturalHeight
  );

  ctx.restore();
};

export const getCroppedImageFile = async (
  image: HTMLImageElement,
  crop: PixelCrop,
  file: File,
  scale = 1,
  rotate = 0
): Promise<File> => {
  try {
    if (!isTestEnv) {
      const canvas = document.createElement('canvas');
      canvasPreview(image, canvas, crop, scale, rotate);

      const blob = await toBlob(canvas, file.type);

      return blob ? new File([blob], file.name, { type: file.type }) : file;
    } else {
      return file;
    }
  } catch (e) {
    return file;
  }
};

export const getCompressedImageFile = async (file: File) => {
  try {
    if (!isTestEnv && COMPRESSABLE_IMAGE_TYPES.includes(file.type)) {
      const blob = await imageCompression(file, {
        maxSizeMB: MAX_IMAGE_SIZE_MB,
        maxWidthOrHeight: MAX_IMAGE_WIDTH,
      });

      return new File([blob], file.name, { type: file.type });
    }
    return file;
  } catch (e) {
    return file;
  }
};
