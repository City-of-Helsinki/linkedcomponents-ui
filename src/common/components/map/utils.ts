import { TFunction } from 'i18next';
import L from 'leaflet';

export const localizeMap = (t: TFunction): void => {
  L.drawLocal.draw.handlers.marker.tooltip.start = t(
    'common.leaflet.drawLocal.draw.handlers.marker.tooltip.start'
  );
  L.drawLocal.draw.toolbar.buttons.marker = t(
    'common.leaflet.drawLocal.draw.toolbar.buttons.marker'
  );
  L.drawLocal.draw.toolbar.actions.text = t(
    'common.leaflet.drawLocal.draw.toolbar.actions.text'
  );
  L.drawLocal.draw.toolbar.actions.title = t(
    'common.leaflet.drawLocal.draw.toolbar.actions.title'
  );
  L.drawLocal.draw.toolbar.finish.text = t(
    'common.leaflet.drawLocal.draw.toolbar.finish.text'
  );
  L.drawLocal.draw.toolbar.finish.title = t(
    'common.leaflet.drawLocal.draw.toolbar.finish.title'
  );
  L.drawLocal.edit.handlers.edit.tooltip.subtext = t(
    'common.leaflet.drawLocal.edit.handlers.edit.tooltip.subtext'
  );
  L.drawLocal.edit.handlers.edit.tooltip.text = t(
    'common.leaflet.drawLocal.edit.handlers.edit.tooltip.text'
  );
  L.drawLocal.edit.handlers.remove.tooltip.text = t(
    'common.leaflet.drawLocal.edit.handlers.remove.tooltip.text'
  );
  L.drawLocal.edit.toolbar.actions.cancel.text = t(
    'common.leaflet.drawLocal.edit.toolbar.actions.cancel.text'
  );
  L.drawLocal.edit.toolbar.actions.cancel.title = t(
    'common.leaflet.drawLocal.edit.toolbar.actions.cancel.title'
  );
  L.drawLocal.edit.toolbar.actions.clearAll.text = t(
    'common.leaflet.drawLocal.edit.toolbar.actions.clearAll.text'
  );
  L.drawLocal.edit.toolbar.actions.clearAll.title = t(
    'common.leaflet.drawLocal.edit.toolbar.actions.clearAll.title'
  );
  L.drawLocal.edit.toolbar.actions.save.text = t(
    'common.leaflet.drawLocal.edit.toolbar.actions.save.text'
  );
  L.drawLocal.edit.toolbar.buttons.edit = t(
    'common.leaflet.drawLocal.edit.toolbar.buttons.edit'
  );
  L.drawLocal.edit.toolbar.buttons.editDisabled = t(
    'common.leaflet.drawLocal.edit.toolbar.buttons.editDisabled'
  );
  L.drawLocal.edit.toolbar.buttons.remove = t(
    'common.leaflet.drawLocal.edit.toolbar.buttons.remove'
  );
  L.drawLocal.edit.toolbar.buttons.removeDisabled = t(
    'common.leaflet.drawLocal.edit.toolbar.buttons.removeDisabled'
  );
};
