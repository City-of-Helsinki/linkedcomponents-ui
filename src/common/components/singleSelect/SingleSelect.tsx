import classNames from 'classnames';
import { Select, SingleSelectProps } from 'hds-react';
import React from 'react';
import { useTranslation } from 'react-i18next';

import { useTheme } from '../../../domain/app/theme/Theme';
import { OptionType } from '../../../types';
import {
  getA11ySelectionMessage,
  getA11yStatusMessage,
} from '../../../utils/accessibilityUtils';
import styles from '../select/select.module.scss';

const SingleSelect: React.FC<SingleSelectProps<OptionType>> = ({
  className,
  ...rest
}) => {
  const { theme } = useTheme();
  const { t } = useTranslation();

  return (
    <Select
      {...rest}
      className={classNames(className, styles.select)}
      getA11yStatusMessage={(options) => getA11yStatusMessage(options, t)}
      getA11ySelectionMessage={
        /* istanbul ignore next */
        (options) => getA11ySelectionMessage(options, t)
      }
      theme={theme.select}
    />
  );
};

export default SingleSelect;
