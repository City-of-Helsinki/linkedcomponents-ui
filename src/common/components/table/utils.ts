import { Header, Order } from './types';

const shouldNotSortRows = (
  order: Order | undefined,
  sorting: string | undefined,
  sortingEnabled: boolean
): boolean => {
  return !sortingEnabled || !order || !sorting;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const defaultSortCompareFunction: (a: any, b: any) => number = (a, b) => {
  if (a < b) {
    return -1;
  }
  if (a > b) {
    return 1;
  }

  return 0;
};

export const processRows = (
  rows: Array<object>,
  order: Order | undefined,
  sorting: string | undefined,
  cols: Header[]
) => {
  const sortingEnabled = cols.some((column) => {
    return column.isSortable === true;
  });

  if (shouldNotSortRows(order, sorting, sortingEnabled)) {
    return [...rows];
  }

  const sortColumn = cols.find((column) => {
    return column.key === sorting;
  });

  const sortCompareFunction =
    sortColumn?.customSortCompareFunction || defaultSortCompareFunction;

  const sortedRows = [...rows].sort((a, b) => {
    const aValue = a[sorting as keyof typeof a];
    const bValue = b[sorting as keyof typeof b];

    return sortCompareFunction(aValue, bValue);
  });

  return order === 'asc' ? sortedRows : sortedRows.reverse();
};
