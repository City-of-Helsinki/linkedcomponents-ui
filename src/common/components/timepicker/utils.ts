import { TimeObject } from './types';

export const formatTime = ({ hours, minutes }: TimeObject): string => {
  const hour = hours.toString().padStart(2, '0');
  const minute = minutes.toString().padStart(2, '0');
  return `${hour}.${minute}`;
};

export const getTimes = (interval: number): string[] => {
  const times = [];

  for (let h = 0; h < 24; h += 1) {
    for (let m = 0; m < 60; m += interval) {
      times.push(formatTime({ hours: h, minutes: m }));
    }
  }

  return times;
};

export const getTimeObjects = (interval: number): TimeObject[] => {
  const times: TimeObject[] = [];

  for (let h = 0; h < 24; h += 1) {
    for (let m = 0; m < 60; m += interval) {
      times.push({ hours: h, minutes: m });
    }
  }

  return times;
};

export const getTimeObject = (time: string): TimeObject => {
  const hours = Number(time.replace(':', '.').split('.')[0]);
  const minutes = Number(time.replace(':', '.').split('.')[1]);

  return {
    hours: Number.isInteger(hours) && hours < 24 && hours >= 0 ? hours : 0,
    minutes:
      Number.isInteger(minutes) && minutes < 60 && minutes >= 0 ? minutes : 0,
  };
};
