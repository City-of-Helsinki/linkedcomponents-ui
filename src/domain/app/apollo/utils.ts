import { EMPTY_MULTI_LANGUAGE_OBJECT } from '../../../constants';
import {
  DataSource,
  Division,
  Enrolment,
  Event,
  ExternalLink,
  Image,
  Keyword,
  KeywordSet,
  Language,
  LocalisedObject,
  Meta,
  Offer,
  Organization,
  OrganizationClass,
  Place,
  Position,
  Registration,
  User,
} from '../../../generated/graphql';
import getValue from '../../../utils/getValue';

export const addTypenameDataSource = (
  dataSource?: DataSource | null
): DataSource | null =>
  dataSource
    ? {
        ...dataSource,
        __typename: 'DataSource',
      }
    : null;

export const addTypenameDivision = (
  division?: Division | null
): Division | null =>
  division
    ? {
        ...division,
        name: addTypenameLocalisedObject(division.name),
        __typename: 'Division',
      }
    : null;

export const addTypenameEnrolment = (
  enrolment?: Enrolment | null
): Enrolment | null =>
  enrolment
    ? {
        ...enrolment,
        __typename: 'Enrolment',
      }
    : null;

export const addTypenameEvent = (event?: Event | null): Event | null =>
  event
    ? {
        ...event,
        description: addTypenameLocalisedObject(event.description),
        infoUrl: addTypenameLocalisedObject(event.infoUrl),
        name: addTypenameLocalisedObject(event.name),
        provider: addTypenameLocalisedObject(event.provider),
        shortDescription: addTypenameLocalisedObject(event.shortDescription),
        externalLinks: getValue(event.externalLinks, [])?.map((externalLink) =>
          addTypenameExternalLink(externalLink)
        ),
        images: getValue(
          event.images?.map((image) => addTypenameImage(image)),
          []
        ),
        inLanguage: getValue(
          event.inLanguage?.map((language) => addTypenameLanguage(language)),
          []
        ),
        keywords: getValue(
          event.keywords?.map((keyword) => addTypenameKeyword(keyword)),
          []
        ),
        audience: getValue(
          event.audience?.map((keyword) => addTypenameKeyword(keyword)),
          []
        ),
        location: addTypenamePlace(event.location),
        offers: getValue(
          event.offers?.map((offer) => addTypenameOffer(offer)),
          []
        ),
        superEvent: event.superEvent
          ? addTypenameEvent(event.superEvent)
          : null,
        subEvents: getValue(
          event.subEvents?.map((subEvent) => addTypenameEvent(subEvent)),
          []
        ),
        __typename: 'Event',
      }
    : null;

export const addTypenameExternalLink = (
  externalLink?: ExternalLink | null
): ExternalLink | null =>
  externalLink ? { ...externalLink, __typename: 'ExternalLink' } : null;

export const addTypenameImage = (image?: Image | null): Image | null =>
  image
    ? {
        ...image,
        altText: addTypenameLocalisedObject(image.altText),
        __typename: 'Image',
      }
    : null;

export const addTypenameKeyword = (keyword?: Keyword | null): Keyword | null =>
  keyword
    ? {
        ...keyword,
        name: addTypenameLocalisedObject(keyword?.name),
        __typename: 'Keyword',
      }
    : null;

export const addTypenameKeywordSet = (
  keywordSet?: KeywordSet | null
): KeywordSet | null =>
  keywordSet
    ? {
        ...keywordSet,
        keywords: getValue(
          keywordSet.keywords?.map((keyword) => addTypenameKeyword(keyword)),
          []
        ),
        name: addTypenameLocalisedObject(keywordSet?.name),
        __typename: 'KeywordSet',
      }
    : null;

export const addTypenameLanguage = (
  language?: Language | null
): Language | null =>
  language
    ? {
        ...language,
        name: addTypenameLocalisedObject(language?.name),
        __typename: 'Language',
      }
    : null;

export const addTypenameLocalisedObject = (
  item?: string | LocalisedObject | null
): LocalisedObject | null =>
  item
    ? {
        ...(typeof item === 'string'
          ? { ...EMPTY_MULTI_LANGUAGE_OBJECT, fi: item }
          : item),
        __typename: 'LocalisedObject',
      }
    : null;

export const addTypenameMeta = (meta: Meta): Meta => ({
  ...meta,
  __typename: 'Meta',
});

export const addTypenameOffer = (offer?: Offer | null): Offer | null =>
  offer
    ? {
        ...offer,
        description: addTypenameLocalisedObject(offer.description),
        infoUrl: addTypenameLocalisedObject(offer.infoUrl),
        price: addTypenameLocalisedObject(offer.price),
        __typename: 'Offer',
      }
    : null;

export const addTypenameOrganization = (
  organization?: Organization | null
): Organization | null =>
  organization ? { ...organization, __typename: 'Organization' } : null;

export const addTypenameOrganizationClass = (
  organizationClass?: OrganizationClass | null
): OrganizationClass | null =>
  organizationClass
    ? { ...organizationClass, __typename: 'OrganizationClass' }
    : null;

export const addTypenamePlace = (place?: Place | null): Place | null =>
  place
    ? {
        ...place,
        addressLocality: addTypenameLocalisedObject(place.addressLocality),
        divisions: getValue(
          place.divisions?.map((division) => addTypenameDivision(division)),
          []
        ),
        infoUrl: addTypenameLocalisedObject(place.infoUrl),
        name: addTypenameLocalisedObject(place.name),
        position: addTypenamePosition(place.position),
        streetAddress: addTypenameLocalisedObject(place.streetAddress),
        telephone: addTypenameLocalisedObject(place.telephone),
        __typename: 'Place',
      }
    : null;

export const addTypenamePosition = (
  position?: Position | null
): Position | null =>
  position ? { ...position, __typename: 'Position' } : null;

export const addTypenameRegistration = (
  registration?: Registration | null
): Registration | null =>
  registration
    ? {
        ...registration,
        event: addTypenameEvent(registration?.event),
        signups: Array.isArray(registration.signups)
          ? registration.signups.map((enrolment) =>
              addTypenameEnrolment(enrolment)
            )
          : [],
        __typename: 'Registration',
      }
    : null;

export const addTypenameUser = (user?: User | null): User | null =>
  user ? { ...user, __typename: 'User' } : null;
