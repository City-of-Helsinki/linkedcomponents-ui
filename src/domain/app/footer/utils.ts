import { ROUTES } from '../../../constants';
import { UserFieldsFragment } from '../../../generated/graphql';
import { areRegistrationRoutesAllowed } from '../../user/permissions';
import { NavigationGroup, NavigationGroupItem } from './types';

const getNavigationGroupItemsInstructions = (
  user?: UserFieldsFragment
): NavigationGroupItem[] => [
  {
    label: 'helpPage.sideNavigation.labelEventsInstructions',
    url: ROUTES.INSTRUCTIONS_EVENTS,
  },
  ...(areRegistrationRoutesAllowed(user)
    ? [
        {
          label: 'helpPage.sideNavigation.labelRegistrationInstructions',
          url: ROUTES.INSTRUCTIONS_REGISTRATION,
        },
      ]
    : []),
  {
    label: 'helpPage.sideNavigation.labelFaq',
    url: ROUTES.INSTRUCTIONS_FAQ,
  },
];

export const getNavigationGroupInstructions = (
  user?: UserFieldsFragment
): NavigationGroup => ({
  heading: 'helpPage.sideNavigation.labelInstructions',
  headingLink: ROUTES.INSTRUCTIONS,
  items: getNavigationGroupItemsInstructions(user),
});
