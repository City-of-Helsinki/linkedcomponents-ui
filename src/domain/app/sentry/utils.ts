import * as Sentry from '@sentry/browser';
import * as H from 'history';

import { UserFieldsFragment } from '../../../generated/graphql';

const reportError = ({
  data,
  location,
  message,
  user,
}: {
  data: { error: Record<string, unknown> } & Record<string, unknown>;
  location: H.Location;
  message: string;
  user?: UserFieldsFragment;
}): string => {
  const { error, ...restData } = data;
  const reportObject = {
    extra: {
      data: {
        ...restData,
        currentUrl: window.location.href,
        location,
        timestamp: new Date(),
        user,
        userAsString: user ? JSON.stringify(user) : '',
        userAgent: navigator.userAgent,
        error,
        errorAsString: JSON.stringify(error),
      },
      level: 'error',
    },
  };

  return Sentry.captureException(message, reportObject);
};

export { reportError };
