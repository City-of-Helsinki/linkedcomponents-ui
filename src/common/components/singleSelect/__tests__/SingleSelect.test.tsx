import React from 'react';

import { render } from '../../../../utils/testUtils';
import SingleSelect from '../SingleSelect';

const options = [
  { label: 'Option 1', value: 'option1' },
  { label: 'Option 2', value: 'option2' },
];

const renderComponent = () =>
  render(<SingleSelect label="Select label" options={options} />);

test('renders the component', async () => {
  const { container } = renderComponent();

  expect(container).toMatchSnapshot();
});
