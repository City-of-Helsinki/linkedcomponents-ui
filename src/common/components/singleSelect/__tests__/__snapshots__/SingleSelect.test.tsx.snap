// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div>
  <div
    class="Select-module_root__Ka5uO custom-theme-2 select"
  >
    <label
      class="FieldLabel-module_label__1zrXK "
      for="hds-select-1-toggle-button"
      id="hds-select-1-label"
    >
      Select label
    </label>
    <div
      class="Select-module_wrapper__1WQXs"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="hds-select-1-label hds-select-1-toggle-button"
        aria-owns="hds-select-1-menu"
        class="Select-module_button__1aIsm Select-module_placeholder__21c0j"
        id="hds-select-1-toggle-button"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik Select-module_angleIcon__2-9AD"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M0 0h24v24H0z"
            />
            <path
              d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
              fill="currentColor"
            />
          </g>
        </svg>
      </button>
      <ul
        aria-labelledby="hds-select-1-label"
        class="Select-module_menu__1H2aU"
        id="hds-select-1-menu"
        role="listbox"
        style="max-height: 260px;"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;
